contributors/design-proposals/local-pv-provisioner.md

```diff
@@ -0,0 +1,160 @@
+
+### Minion Provisioner
+The minion provisioner retains most key components from the previous design, including the Cache, Discoverer, and Deleter. The Populator is moved to master because it interacts with the API server. In order to interact with the master, it exposes the following gRPC calls:
+
+| RPC                      | Description |
+| --- | --- |
+| `Add/Update/RemovePV(PersistentVolume)` | Updates the Cache with the given PV. Called by master when PV informer is triggered. |
+| `PushCurrentPVs([]PersistentVolume)`    | Merges given list of PVs with the Cache. Used by the Populator. |
+| `GetPVRequest()`                        | Peeks at the next unhandled request in the queue. |
+| `CompletePVRequest(PVRequest)`          | Signals that the last request is completed and can be removed from the queue. |
+
+In the case of master failure while a request is being processed, the minion needs to keep record of the request so that a new master can process it again. Thus it's necessary for master to signal request completion.
+
+### Master Provisioner
+The populator is initialized so that it calls the appropriate gRPC for every PV Informer update. Then it populates each minion's cache by calling `PushCurrentPVs()`, passing the collection of PVs to be controlled by this minion.
+
+The master then performs minion discovery as described in the [Minion Discovery](#minion-discovery) section of design 1, and spawns a goroutine for each minion. Each goroutine periodically calls `GetPVRequest()`, performing that request if it receives one. Then it calls `CompletePVRequest()` to signal the request completion.
+
+## Design 3: Event-Driven with Custom Resources
+
+### High Level Overview
+The previous two designs are complicated by the asynchronous nature of delete and the presence of failures. 
+As recommended by @dhirajh, another design is to persist ProvisionerStatus as a Custom Resource, and to drive all operations with PV and ProvisionerStatus Informer events.
+
+To ensure no malicious pods can affect provisioner execution, the CustomResourceDefinition and all pods must be in a trusted admin namespace.
+
+In the following description, we assume all Informer updates are cached and events buffered. On initialization, all caches are populated, and every Informer update event handler is executed once to ensure the entire provisioner system is in a consistent state.
+
+### ProvisionerStatus CRD
+The CRD contains the following fields (similar to design 1):
+
+* `LocalVolumes` - A collection of volumes mounted in the local storage directory on the node. The discoverer is triggered to acquire this information.
+* `DeleteOperations` - A list keeping track of information related to each delete operation. This status is managed by the cleaner. It contains the following fields:
```

Comment:
I may be missing something obvious, but why we need to keep track of DeleteOperations at all? Provisioner/minion can get the signal to clean a local volume by watching PVs. Or is reading/watching all PVs on API server so dangerous?

Status of the cleanup is IMO internal state of the minion, do we need it to be exposed? The cleanup is implicitly "in progress" when PV is missing for a local volume. When it's complete, corresponding PV is created.

Reply:
The problem is that the user can prematurely delete the PV object before it's done cleaning.  If the minion restarts, it loses the in memory state of the cleaning status.

However if we assume PV deletion issue is going to be fixed, then that would simplify this design and remove the need for a crd.

Reply:
The minion can have some local storage to keep track of cleanups in progress.

Reply:
If the minion has to keep track of state anyway, one advantage of keeping state in the CRD is visibility to admins.

But in any case, giving minions write access to CRDs is no longer an option because of CRD isolation concerns, so each minion has to store local state anyway.

Reply:
I actually think the minion need not maintain any state locally. If we follow the pattern mentioned above of pinging the master when a state needs to be updated (and having the master call back/authenticate the minon), then the CRD effectively acts as the persistent store for the minion. I certainly like the ability CRD provides for fetching status of cleaning for all disks via kubectl. The other option of logging into the host and poking around local files and making sense of state in them seems less appealing.

Reply:
Can someone just query the PVs to get the deletion state?  We're going to log an event in the PV when deletion is started and if deletion failed.

Reply:
@dhirajh CRD becomes a less reliable persistent store due to an unreliable network and possibility of master crashes. In this case we're probably OK as long as the delete request is persisted.